File: Simulation/Core/Solver.cs
using System.Collections.Generic;
using System.Linq;

namespace HistoryTeller.Simulation
{
    /// <summary>
    /// Brute-force солвер: проверяет решаемость уровня перебором всех расстановок.
    /// Для контент-пайплайна и подсказок; пространство поиска у таких пазлов маленькое.
    /// </summary>
    public static class Solver
    {
        public sealed class Result
        {
            public List<List<Panel>> Solutions = new List<List<Panel>>();
            public long SearchSpace;
            public bool IsSolvable => Solutions.Count > 0;
        }

        public static Result Solve(LevelDef level, ContentDb db, int maxSolutions = int.MaxValue)
        {
            var options = PanelOptions(level, db);
            var result = new Result
            {
                SearchSpace = Pow(options.Count, level.Panels)
            };

            foreach (var assignment in CartesianPower(options, level.Panels))
            {
                var world = Engine.Simulate(assignment, db, null, level.CreateInitialWorld());
                if (level.Goal.IsMet(world))
                {
                    result.Solutions.Add(assignment);
                    if (result.Solutions.Count >= maxSolutions) break;
                }
            }
            return result;
        }

        /// <summary>Все варианты заполнения панели. Порядок персонажей = роли-слоты → размещения.</summary>
        private static List<Panel> PanelOptions(LevelDef level, ContentDb db)
        {
            var options = new List<Panel>();
            foreach (var sceneId in level.Scenes)
            {
                int slots = db.Scenes[sceneId].Slots;
                int maxK = System.Math.Min(slots, level.Characters.Count);
                for (int k = 0; k <= maxK; k++)
                    foreach (var perm in Permutations(level.Characters, k))
                        options.Add(new Panel { SceneId = sceneId, Characters = perm });
            }
            return options;
        }

        /// <summary>Упорядоченные размещения длины k.</summary>
        private static IEnumerable<List<string>> Permutations(List<string> items, int k)
        {
            if (k == 0) { yield return new List<string>(); yield break; }
            if (k > items.Count) yield break;
            for (int i = 0; i < items.Count; i++)
            {
                var rest = new List<string>(items);
                rest.RemoveAt(i);
                foreach (var tail in Permutations(rest, k - 1))
                {
                    tail.Insert(0, items[i]);
                    yield return tail;
                }
            }
        }

        private static IEnumerable<List<Panel>> CartesianPower(List<Panel> options, int n)
        {
            var indices = new int[n];
            while (true)
            {
                yield return indices.Select(i => options[i]).ToList();
                int pos = n - 1;
                while (pos >= 0 && ++indices[pos] == options.Count)
                    indices[pos--] = 0;
                if (pos < 0) yield break;
            }
        }

        private static IEnumerable<List<string>> Combinations(List<string> items, int k)
        {
            if (k == 0) { yield return new List<string>(); yield break; }
            for (int i = 0; i <= items.Count - k; i++)
                foreach (var tail in Combinations(items.Skip(i + 1).ToList(), k - 1))
                {
                    tail.Insert(0, items[i]);
                    yield return tail;
                }
        }

        private static long Pow(long b, int e)
        {
            long r = 1;
            for (int i = 0; i < e; i++) r *= b;
            return r;
        }
    }
}

File: ios/Modules/Simulation/Sources/Simulation/Solver.swift
import Foundation

/// Brute-force солвер: проверяет решаемость уровня перебором всех расстановок.
/// Для контент-пайплайна и подсказок; пространство поиска у таких пазлов маленькое.
public enum Solver {

    public struct Result {
        public var solutions: [[Panel]] = []
        public var searchSpace: Int = 0
        public var isSolvable: Bool { !solutions.isEmpty }
    }

    public static func solve(_ level: LevelDef, _ db: ContentDb, maxSolutions: Int = .max) -> Result {
        let options = panelOptions(level, db)
        var result = Result()
        result.searchSpace = pow(options.count, level.panels)
        if options.isEmpty { return result }

        let n = level.panels
        var indices = [Int](repeating: 0, count: n)
        // Одометр по индексам опций — ленивый декартов степенной перебор (без материализации всех расстановок).
        while true {
            let assignment = indices.map { options[$0] }
            let world = Engine.simulate(assignment, db, initial: level.createInitialWorld()).world
            if level.goal.isMet(world) {
                result.solutions.append(assignment)
                if result.solutions.count >= maxSolutions { break }
            }
            var pos = n - 1
            while pos >= 0 {
                indices[pos] += 1
                if indices[pos] == options.count { indices[pos] = 0; pos -= 1 }
                else { break }
            }
            if pos < 0 { break }
        }
        return result
    }

    /// Все варианты заполнения панели. Порядок персонажей = роли-слоты → размещения.
    static func panelOptions(_ level: LevelDef, _ db: ContentDb) -> [Panel] {
        var options: [Panel] = []
        for sceneId in level.scenes {
            let slots = db.scenes[sceneId]!.slots
            let maxK = min(slots, level.characters.count)
            for k in 0...maxK {
                for perm in Engine.permutations(level.characters, k) {
                    options.append(Panel(sceneId: sceneId, characters: perm))
                }
            }
        }
        return options
    }

    static func pow(_ base: Int, _ exp: Int) -> Int {
        var r = 1
        for _ in 0..<exp { r *= base }
        return r
    }
}
